File: src/services/folders.ts
import { invoke } from "@tauri-apps/api/core";
import { ChatFolder, Message, UpdateChatAddedToFolder, UpdateChatRemovedFromFolder } from "../model";
import { listen } from "@tauri-apps/api/event";
import { store } from "../store";
import { getDefaultChatFolder } from "../settings/settings";

export async function handleFoldersUpdates() {
  return [
    await listen<ChatFolder[]>('update-chat-folders', async (event) => {
      store.chatFolders = event.payload;
      const defaultChatFolder = await getDefaultChatFolder();
      store.selectedChatFolderId = store.chatFolders.find(f => f.id === defaultChatFolder)?.id || 0;
    }),
    await listen<UpdateChatAddedToFolder>('update-chat-added-to-folder', (event) => {
      const update = event.payload;
      store.addChatToFolder(update.folder_id, update.chat_id);
    }),
    await listen<UpdateChatRemovedFromFolder>('update-chat-removed-from-folder', async (event) => {
      const update = event.payload;
      store.removeChatFromFolder(update.folder_id, update.chat_id);
      if (update.folder_id === 0) {
        store.deleteChat(update.chat_id);
      }
    }),
  ]
}

export async function createFolder(title: string, includedChatIds: number[]) {
  await invoke<Message>('create_chat_folder', { title, includedChatIds })
}

export async function deleteFolder(folderId: number) {
  await invoke<Message>('delete_chat_folder', { folderId })
}

export async function renameFolder(folderId: number, newTitle: string) {
  await invoke<Message>('rename_chat_folder', { folderId, newTitle })
}

export async function addChatToFolder(chatId: number, folderId: number) {
  await invoke<Message>('add_chat_to_folder', { chatId, folderId })
}

export async function removeChatFromFolder(chatId: number, folderId: number) {
  await invoke<Message>('remove_chat_from_folder', { chatId, folderId })
}

File: src-tauri/src/telegram/folders.rs
use crate::{
    emit,
    model::{ChatFolder, UpdateChatAddedToFolder, UpdateChatRemovedFromFolder},
};

use tdlib::enums::{ChatList, Update};

pub async fn handle_folders_update<R: tauri::Runtime>(
    app: &tauri::AppHandle<R>,
    update: &Update,
) -> bool {
    match update {
        Update::ChatFolders(update) => {
            let mut folders = vec![ChatFolder::new(0, String::from("Main"))];
            for folder in &update.chat_folders {
                folders.push(ChatFolder::from(&folder));
            }
            emit(app, "update-chat-folders", folders);
            return true;
        }
        Update::ChatAddedToList(update) => {
            if update.chat_list == ChatList::Main {
                emit(
                    app,
                    "update-chat-added-to-folder",
                    UpdateChatAddedToFolder {
                        chat_id: update.chat_id,
                        folder_id: 0,
                    },
                );
            } else if let tdlib::enums::ChatList::Folder(folder) = &update.chat_list {
                emit(
                    app,
                    "update-chat-added-to-folder",
                    UpdateChatAddedToFolder {
                        chat_id: update.chat_id,
                        folder_id: folder.chat_folder_id,
                    },
                );
            }
            return true;
        }
        Update::ChatRemovedFromList(update) => {
            if update.chat_list == ChatList::Main {
                emit(
                    app,
                    "update-chat-removed-from-folder",
                    UpdateChatRemovedFromFolder {
                        chat_id: update.chat_id,
                        folder_id: 0,
                    },
                );
            } else if let tdlib::enums::ChatList::Folder(folder) = &update.chat_list {
                emit(
                    app,
                    "update-chat-removed-from-folder",
                    UpdateChatRemovedFromFolder {
                        chat_id: update.chat_id,
                        folder_id: folder.chat_folder_id,
                    },
                );
            }
            return true;
        }
        _ => {
            return false;
        }
    }
}
